Omindex libextractor

Exter-dg opened the pull request:
Adds support for Audio and Video formats in Omega (Extracting meta-data).

ojwb commented:
> I noticed that if the global variables are not cleared/resized to 0, the values from previous file stays and is added to the new file too.
>
> Why does this happen?

Well, that's how global variables work - they have the lifetime of the program (in this case of the handler program).

> Is there a better alternative to global variables here?

Yes - the `cls` parameter to `EXTRACTOR_extract()` is provided to address this need.  You can make a class or structure that holds pointers or references to the `title`, etc parameters and pass a pointer to that structure in as `cls`.   Then cast that back to the right type in your callback function.  It's a pretty common idiom that a callback function takes a `void*` pointer so you can pass in a pointer to something in this way.

Note that you really want to pointers or references in this class/structure to avoid copying these strings needlessly (or else I guess you could use `std::move()` - the point is that these are potentially long strings so we want to avoid the overhead of unnecessary copying of them).

ojwb commented on xapian-applications/omega/configure.ac:
```diff
@@ -286,6 +286,11 @@ PKG_CHECK_MODULES([LIBARCHIVE], [libarchive], [
+dnl check if libextractor is available.
+PKG_CHECK_MODULES([LIBEXTRACTOR], [libextractor], [
+AC_DEFINE(HAVE_LIBEXTRACTOR, 1, [Define HAVE_LIBEXTRACTOR if the library is available])
+OMINDEX_MODULES="$OMINDEX_MODULES omindex_libextractor"],[ ])
```
This isn't indented like the similar constructs just above.

Exter-dg replied:
I have changed this.

barufa commented on xapian-applications/omega/handler_libextractor.cc:
```diff
@@ -0,0 +1,120 @@
+	::keywords.resize(0);
+	::title.resize(0);
+	::author.resize(0);
+	::pages.resize(0);
```
A better option than resizing the string to zero would be calling the string method [clear](http://www.cplusplus.com/reference/string/string/clear/).

Exter-dg replied:
Thanks for pointing this out. I was using clear(), I chose to use resize(0) because I read that clear (in case of vector) will make the vector empty but space will still be reserved, So I used resize instead of clear. I am not sure how does it work in case of strings. I will research more about this.
For the handler - I have used a struct as suggested by Olly.

ojwb replied:
https://en.cppreference.com/w/cpp/string/basic_string/clear says:

> Unlike for std::vector::clear, the C++ standard does not explicitly require that capacity is unchanged by this function, but existing implementations do not change capacity. This means that they do not release the allocated memory (see also shrink_to_fit). 

(I'd definitely recommends cppreference.com as a resource - I think there's an offline version too.)

But if you're reusing the strings over and over you probably want the reserved space to be kept.  It's not so helpful if you process one enormous file but mostly small ones, but it's also not helpful to be constantly releasing and reallocating space.

barufa commented on xapian-applications/omega/handler_libextractor.cc:
```diff
@@ -0,0 +1,120 @@
+
+using namespace std;
+
+string title, keywords, author, pages;
```
Not sure if using the same names for this variables is a good option as this practice can induce bugs. I think we should try to avoid it.

We can use an structure(as Olly've suggested) to store the necessary information and pass a pointer to it in `cls` instead of global variables.

Exter-dg replied:
I have changed this.

Exter-dg replied:
Should I avoid using the same variables inside the structure too? As I have done right now.

barufa replied:
No, that's not necessary. Only when two variables have the same name under a scope. This is not an error, but it can introduce bugs in the code (for example, if you forget :: when you want to refer to a global variable) and makes the code less readable.

barufa commented on xapian-applications/omega/handler_libextractor.cc:
```diff
@@ -0,0 +1,120 @@
+	case EXTRACTOR_METATYPE_TITLE:
+	case EXTRACTOR_METATYPE_BOOK_TITLE:
+	case EXTRACTOR_METATYPE_SUBJECT:
+	    title = data;
```
I couldn't find if `data` is null terminated on libextractor docs. Maybe using `data_len` would be a good idea.
There is actually a [string constructor](http://www.cplusplus.com/reference/string/string/string/) that uses it(see constructor number 5).

Exter-dg replied:
Thanks for pointing this out. I have changed to code and now it uses the string constructor 

`string (const char* s, size_t n);`

ojwb commented on xapian-applications/omega/handler_libextractor.cc:
```diff
@@ -0,0 +1,133 @@
+	case EXTRACTOR_METATYPE_SUBJECT:
+	case EXTRACTOR_METATYPE_TITLE:
+	    md->title += string(data, data_len);
+	    md->title += " ";
```
This creates a temporary `std::string` object to append the data, which will result in needlessly copying the data, and also in a heap allocation if the string is at all long (short strings are typically actually just stored within the `std::string` object itself).  This temporary object can be avoided by using the `std::string::append()` method - that has an overload form which takes a pointer and length:

```
md->title.append(data, data_len);
```

Also, this always results in a trailing space on the result - it'd be neater to only add it if there's already something in the variable, which is easy to do by first checking and adding the space if non-empty, then appending the data.

Exter-dg replied:
I have made the necessary changes here.

ojwb commented on xapian-applications/omega/handler_libextractor.cc:
```diff
@@ -74,12 +74,13 @@ process_metadata(void* cls,
 
 	case EXTRACTOR_METATYPE_PAGE_COUNT:
-	    md->pages = string(data, data_len);
+	    md->pages.append(data, data_len);
```
If we get multiple `EXTRACTOR_METATYPE_PAGE_COUNT` items (which seems like it would happen if more than one plugin reports a number of pages for a format) then this will now concatenate them as strings - e.g. if we're told 12 pages twice, we'll turn that into `1212`!

The behaviour before of taking the last answer in case of duplicates seems better, so I think you want the "assign" method instead of "append" here.

Exter-dg replied:
Thanks for pointing this out. I have changed this.

ojwb commented on xapian-applications/omega/mimemap.tokens:
```diff
@@ -214,6 +214,23 @@ tif		image/tiff
+# Audio files:
+flac		audio/flac
+mp3		audio/mpeg
+ogg		audio/ogg
```
There's also `oga` and `spx` for `audio/ogg`, and `ogv` for `video/ogg` - see https://tools.ietf.org/html/rfc5334#section-10.2

I think we probably don't want an extension to mime-type mapping for the `ogg` extension itself though - `ogg` can be `audio/ogg` or `video/ogg` (as your example files demonstrate!)  If we don't provide a mapping then omindex will use libmagic which can tell the difference from the file content:

```
$ file --mime-type testfiles/*/*.ogg
testfiles/audio/file_example_OOG_1MG.ogg:       audio/ogg
testfiles/video/file_example_OGG_480_1_7mg.ogg: video/ogg
```

(`file` uses libmagic internally.)

Both get mapped to libextractor by default, but the recorded mime-type will be wrong (which could matter if that's provided as a "file type" filter to users) and it'll also be confusing to the user if they overrides the default mappings.

Exter-dg replied:
Thank you for the insight, I have changed this.

ojwb replied:
You've addressed the second part, but not the missing extensions I noted in the first paragraph.

We should pick up the type of such files via libmagic, but it's good to provide extension to mime type mappings where they're reliable since libmagic needs to look at the start of the file to determine the type, which takes more time than just looking at the end of the filename - not much per file, but it can add up.

Exter-dg replied:
Oh, I misunderstood the comment. I will fix this and update the PR.

Exter-dg replied:
I have fixed this

ojwb commented on xapian-applications/omega/omindexcheck.cc:
```diff
@@ -170,8 +181,14 @@ main(int argc, char** argv)
+		startswith(iter->first, "video"))
+		skip &= compare_test(iter->second, doc, url);
+	    else
+		succeed &= compare_test(iter->second, doc, url);
```
I think we need to find a better approach than this.

As written, the test can never fail for an audio or video file, but there are test results which are clearly a failure because we should either get all the terms that come from a plugin or none of them.  So for example, for `audio/file_example_OOG_1MG.ogg` if we get `Akevin` we should also get `Simpact`.

You need to think carefully about what cases might happen due to a missing plugin and what can't, so that the test can actually catch problems.

Also, the variable `skip` seems to be confusingly named - it looks to me like `skip` being `false` means we're skipping the tests if they'd otherwise pass.  The sense seems reversed from what I'd expect.

(And a coding style point - it's clearer to put `{`...`}` on the blocks for an `if` where the condition has been split over multiple lines like this.)

Exter-dg commented on xapian-applications/omega/Makefile.am:
```diff
@@ -44,7 +44,7 @@ BUILT_SOURCES = extra/omegascript.vim my-html-tok.h mimemap.h namedents.h
 MAINTAINERCLEANFILES = my-html-tok.h mimemap.h mimemap.stamp namedents.h
 
-EXTRA_PROGRAMS = omindex_poppler omindex_libebook omindex_libetonyek omindex_tesseract omindex_gmime omindex_libarchive omindex_libabw omindex_libcdr
+EXTRA_PROGRAMS = omindex_poppler omindex_libebook omindex_libetonyek omindex_tesseract omindex_gmime omindex_libarchive omindex_libabw omindex_libcdr omindex_libextractor
```
I have changed this.

Exter-dg commented on xapian-applications/omega/omindexcheck.cc:
```diff
@@ -34,6 +34,8 @@ typedef vector<string> testcase;
 
 static unordered_map<string, testcase> tests;
 
+enum err_code { PASS, FAIL, SKIP };
```
Thank you for pointing this out. I have changed the name to `test_result`.

Exter-dg commented on xapian-applications/omega/omindexcheck.cc:
```diff
@@ -92,8 +94,8 @@ index_test()
-    // blank file
-    tests.insert({"odf/blank.odt", {}});
+    // blank file - will skip the test as no terms will be found.
+    // tests.insert({"odf/blank.odt", {}});
```
I have made the necessary changes in a new commit.

Exter-dg commented on xapian-applications/omega/omindexcheck.cc:
```diff
@@ -157,29 +159,54 @@ index_test()
 }
 
-static bool
+static err_code
```
Thank you for your insight. I will try and make separate commits for distinct changes in the future.

Exter-dg commented on xapian-applications/omega/omindexcheck.cc:
```diff
@@ -157,29 +159,54 @@ index_test()
-	    return false;
+	    exists = false;
+	} else if (!term_found) {
+	    term_found = true;
```
Thank you for pointing this out. I have changed the `else if` to `else`.

Exter-dg commented on xapian-applications/omega/omindexcheck.cc:
```diff
@@ -157,29 +159,54 @@ index_test()
-    return true;
+    // Skips if no term is found
+    // Passes if all terms are found
+    // Fails if only some terms are found
```
I have changed this. The comments are now at the top of the function.

Exter-dg commented on xapian-applications/omega/omindexcheck.cc:
```diff
@@ -157,29 +159,54 @@ index_test()
 {
     Xapian::Database db;
-    bool succeed = true;
+    err_code succeed = PASS;
```
Thank you for pointing this out. Right now, I have changed 2 variable names - `result` will indicate the result of the entire test and `outcome` will indicate the result of individual tests on files. Are these better names?

* result -> outcome
* succeed -> result

Exter-dg commented on xapian-applications/omega/omindexcheck.cc:
```diff
@@ -196,8 +223,17 @@ main(int argc, char** argv)
+    else if (succeed == FAIL)
+	return 1;
+    else
+	return 77;
```
Thank you for pointing this out. I have added a comment for the same.

ojwb commented on xapian-applications/omega/handler_libextractor.cc:
```diff
@@ -0,0 +1,136 @@
+* data_mime_type: mimetype of data according to libextractor
+* data: actual meta-data found
+* data_len: number of bytes in data
+*/
```
For Xapian code, we'd usually indent the `*` and `*/` lines so the `*`s all line up vertically with the one in `/*`.

It's also better to format function parameter docs as a doxygen documentation comment - that way someone can run doxygen over the code and get collated docs for internal APIs.  Going through and fixing up existing docs that lack such formatting is tedious work which is hard to automate, so it's the sort of thing which doesn't ever get done, so doing this when the comments are added is better.

What that actually means is starting the comment block with `/**` followed by a brief description of the purpose of the function, and then formatting the parameter list entries like: `@param data_len  number of bytes in data`

Exter-dg replied:
I have made these changes and updated the comment.

ojwb commented on xapian-applications/omega/handler_libextractor.cc:
```diff
@@ -0,0 +1,136 @@
+		 const char* data,
+		 size_t data_len)
+{
+    struct metadata* md = (struct metadata*)cls;
```
Better to avoid C-style casts in C++ code.  A C-style cast can be used for one of several purposes and it may not be clear to the reader which are wanted.  Also the compiler doesn't know which, and that allows errors to slip through.  C++ casts make the purpose clear to the reader, and the compiler can complain if they're misused.  `static_cast<>` is the appropriate one here: https://en.cppreference.com/w/cpp/language/static_cast

As an example of the sort of problem this avoids, if we had `const void* cls2` and tried:

```
    struct metadata* md = (struct metadata*)cls2;
```

Then this would compile without even a warning (because one use of a C-style cast is to remove `const` qualification), and we'd be able to modify the object despite getting a `const` pointer to it - that's potentially very bad.  If the object is actually in the `RODATA` section, it could give an access violation.  If it's constant data but not protected in that way, it's undefined behaviour to modify it.  Even if it's not constant, someone reading code that passes in an object as a `const` pointer won't expect it to be modified, and reasoning about calling code may be invalid.

If you try to compile this you'll get an error:

```
    struct metadata* md = static_cast<struct metadata*>(cls2);
```

If you actually want to remove `const` with a C++ cast you need to use `const_cast<>` (and that only removes `const` - it won't accidentally change the type at the same time).

Exter-dg replied:
Thank you for your insight. I have changed this.

ojwb commented on xapian-applications/omega/handler_libextractor.cc:
```diff
@@ -0,0 +1,136 @@
+
+	// Add all default plugins
+	struct EXTRACTOR_PluginList* plugins
+	 = EXTRACTOR_plugin_add_defaults(EXTRACTOR_OPTION_DEFAULT_POLICY);
```
Xapian coding style is to wrap after `=` not before.

But you can probably avoid needing to wrap by using `auto` for the type.  The exact type isn't interesting here - this is really just an opaque thing representing plugins which we need to pass to other libextractor API functions.

Exter-dg replied:
I have changed the type to `auto`. But I had to still wrap the lines as it was still exceeding 80 chars.

ojwb commented on xapian-applications/omega/handler_libextractor.cc:
```diff
@@ -0,0 +1,136 @@
+			  &process_metadata, &md);
+	EXTRACTOR_plugin_remove_all(plugins);
+
+    } catch (...) {
```
Blank line at end of block.

Exter-dg replied:
I have corrected this.

ojwb commented on xapian-applications/omega/omindexcheck.cc:
```diff
@@ -92,8 +94,6 @@ index_test()
 #endif
 #if defined HAVE_LIBARCHIVE
-    // blank file
-    tests.insert({"odf/blank.odt", {}});
```
This removes a testcase, which seems undesirable.

Adding per-testcase flags would address this in a much more satisfactory way as it would allow marking the testcases where "skip for no terms" is appropriate.  Then we could keep this testcase, and we'd also catch "missing all terms" bugs for formats where this missing plugin situation doesn't apply (which is probably at least everything that doesn't use libextractor, and perhaps even everything except libextractor with particular plugins).

barufa replied:
Having a flag seems something really useful, but perhaps it would be better making that modification on a new PR (to avoid this become even larger).

Exter-dg replied:
I have made the necessary changes here.
* Change implementation of testcase to a structure.
* Added an enum for flags with 3 values as of now - `FAIL_IF_NO_TERMS`, `SKIP_IF_NO_TERMS`, `PASS_IF_NO_TERMS`

ojwb replied:
I don't think we need 3 flags for this.  If there are 3 flags, there are 8 possible combinations of them, but most of the combinations don't make any sense.  As a general design point, it's cleaner to have a design where there aren't invalid combinations like this to worry about.

I think there are really just 2 cases here:

* We're using an extractor plugin which may not extract anything due to a missing dependency - this is `SKIP_IF_NO_TERMS`
* We aren't - not specifying `SKIP_IF_NO_TERMS` covers this case (and if you add a constructor to `struct testcase` with the parameter for `flags` last and with a default of 0 then this you should be able to only specify the flag on the testcases that need it and avoid the clutter of flags on every testcase in the list).

Note that you can tell the cases that you currently have marked with `FAIL_IF_NO_TERMS` and `PASS_IF_NO_TERMS` apart because the latter has `terms.empty()` and the former don't.  It wouldn't really make sense to list terms but have `PASS_IF_NO_TERMS`, and a testcase listing no terms with `FAIL_IF_NO_TERMS` would have to always fail.

Exter-dg replied:
I have changed this and have updated the PR.

barufa commented:
It seems that the result `omindextest` is `SKIP` in all the environment of Travis, Wouldn't it be a good idea to add the necessary plugins to get at least one `PASS`?

jaylett commented on xapian-applications/omega/omindexcheck.cc:
```diff
@@ -30,156 +30,217 @@
+    sort(test.terms.begin(), test.terms.end());
     Xapian::TermIterator term_iterator = doc.termlist_begin();
-    for (auto& t : test) {
+    bool term_found = false, exists = true;
```
`all_terms_exist` would be slightly better.

Exter-dg replied:
I have changed the variable name to `all_terms_exist`

jaylett commented on xapian-applications/omega/omindexcheck.cc:
```diff
@@ -30,156 +30,217 @@
 
-typedef vector<string> testcase;
+enum test_result { PASS, FAIL, SKIP };
+enum flag { FAIL_IF_NO_TERMS = 1, SKIP_IF_NO_TERMS = 2, PASS_IF_NO_TERMS = 4 };
```
`test_flag` would be one option, or you could nest it in the `struct testcase` which might be more appropriate.

Exter-dg replied:
Thank you for pointing this out. I have changed the variable name to `test_flag`.
I haven't nested the declaration inside the struct as I was not sure if I understood it correctly.

@jaylett  Should I declare the enum `test_flag` inside the structure definition and then the insertion in tests would be of the form-
`tests.insert({"test.txt", {testcase::FAIL_IF_NO_TERMS, {"Zjoey"}}})`

jaylett commented on xapian-applications/omega/omindexcheck.cc:
```diff
@@ -196,8 +257,18 @@ main(int argc, char** argv)
 	auto iter = tests.find(url);
 	if (iter != tests.end()) {
-	    succeed &= compare_test(iter->second, doc, url);
+	    test_result outcome = compare_test(iter->second, doc, url);
```
I'd probably call this `individual_result` or something. It's not going to be clear to someone just reading the variable names what the difference between "outcome" and "result" is.

Exter-dg replied:
Thank you for pointing this out. I have changed the variable name to `individual_result` instead of `outcome`.

Exter-dg commented:
> It seems that the result `omindextest` is `SKIP` in all the environment of Travis, Wouldn't it be a good idea to add the necessary plugins to get at least one `PASS`?

We test 3 formats for libextractor - `avi`, `ogg`, `wav`. For `ogg` libvorbis is a dependency and I have added it to one of the distro in Travis CI. `avi` and `wav` plugins seem to use their own parsers and don't require any additional libraries. The test should pass now. I have updated the PR.
